File: python/src/rag/rag_answer.py
import re
from typing import Any

from common.config import Settings
from providers.openai.generate import generate


def build_prompt(query: str, chunks: list[dict[str, Any]]) -> str:
    context_lines = []
    for chunk in chunks:
        tag = f"[{chunk['source']}:{chunk['id']}]"
        context_lines.append(f"{tag} {chunk['text']}")
    context_block = "\n\n".join(context_lines)

    return (
        "Answer the question using only the context below. "
        "Cite sources in-line as [source:id]. If the answer is not in the context, say you do not know.\n\n"
        f"Context:\n{context_block}\n\n"
        f"Question: {query}\n"
        "Answer:"
    )


def extract_citations(answer: str) -> list[str]:
    citations = re.findall(r"\[([^\[\]]+?:\d+)\]", answer)
    return sorted(set(citations))


def answer_with_citations(
    query: str,
    chunks: list[dict[str, Any]],
    settings: Settings,
) -> tuple[str, list[str]]:
    prompt = build_prompt(query, chunks)
    answer = generate(prompt, settings)
    citations = extract_citations(answer)
    return answer, citations

File: node/src/rag/rag_answer.mjs
import { generate } from "../providers/openai/generate.mjs";

const CITATION_PATTERN = /\[([^\[\]]+?:\d+)\]/g;

// answerWithCitations builds a prompt and extracts inline citations.
export async function answerWithCitations(query, chunks, settings) {
  const prompt = buildPrompt(query, chunks);
  const answer = await generate(prompt, settings);
  const citations = extractCitations(answer);
  return { answer, citations };
}

function buildPrompt(query, chunks) {
  const contextLines = chunks.map((chunk) => {
    const tag = `[${chunk.source}:${chunk.id}]`;
    return `${tag} ${chunk.text}`;
  });
  const contextBlock = contextLines.join("\n\n");

  return (
    "Answer the question using only the context below. " +
    "Cite sources in-line as [source:id]. If the answer is not in the context, say you do not know.\n\n" +
    `Context:\n${contextBlock}\n\n` +
    `Question: ${query}\n` +
    "Answer:"
  );
}

function extractCitations(answer) {
  const citations = new Set();
  let match = CITATION_PATTERN.exec(answer);
  while (match) {
    citations.add(match[1]);
    match = CITATION_PATTERN.exec(answer);
  }
  return Array.from(citations).sort();
}